app/src/main/java/com/duckduckgo/app/sitepermissions/SitePermissionsViewModel.kt

```diff
@@ -0,0 +1,170 @@
+            askCameraEnabled = sitePermissionsRepository.askCameraEnabled,
+            askMicEnabled = sitePermissionsRepository.askMicEnabled
+        )
+    }
+
+    fun allowedSites() {
+        viewModelScope.launch {
+            locationPermissionsRepository.getLocationPermissionsFlow().collect { locationPermissionsList ->
+                sitePermissionsRepository.sitePermissionsWebsitesFlow().collect { sitePermissionsList ->
+                    _viewState.emit(
+                        _viewState.value.copy(
+                            sitesPermissionsAllowed = sitePermissionsList,
+                            locationPermissionsAllowed = locationPermissionsList
+                        )
+                    )
+                }
+            }
+        }
+    }
+
+    fun permissionToggleSelected(
+        isChecked: Boolean,
+        textRes: Int
+    ) {
+        when (textRes) {
+            R.string.sitePermissionsSettingsLocation -> {
+                settingsDataStore.appLocationPermission = isChecked
+                _viewState.value = _viewState.value.copy(askLocationEnabled = isChecked)
+                removeLocationSites()
+                fireTogglePixel(isChecked, PixelValue.LOCATION)
+            }
+            R.string.sitePermissionsSettingsCamera -> {
+                sitePermissionsRepository.askCameraEnabled = isChecked
+                _viewState.value = _viewState.value.copy(askCameraEnabled = isChecked)
+                fireTogglePixel(isChecked, PixelValue.CAMERA)
+            }
+            R.string.sitePermissionsSettingsMicrophone -> {
+                sitePermissionsRepository.askMicEnabled = isChecked
+                _viewState.value = _viewState.value.copy(askMicEnabled = isChecked)
+                fireTogglePixel(isChecked, PixelValue.MIC)
+            }
+        }
+    }
+
+    private fun removeLocationSites() {
+        viewModelScope.launch {
+            geolocationPermissions.clearAll()
+            _viewState.emit(_viewState.value.copy(locationPermissionsAllowed = listOf()))
+        }
+    }
+
+    private fun fireTogglePixel(enabled: Boolean, paramValue: String) {
+        val pixelName = when (enabled) {
+            true -> SitePermissionsPixelName.SITE_PERMISSIONS_ASK_ENABLED
+            false -> SitePermissionsPixelName.SITE_PERMISSIONS_ASK_DISABLED
+        }
+        pixel.fire(pixelName, mapOf(PixelParameter.SITE_PERMISSION to paramValue))
+    }
+
+    fun allowedSiteSelected(domain: String) {
+        viewModelScope.launch {
+            _commands.send(LaunchWebsiteAllowed(domain))
+        }
+    }
+
+    fun removeAllSitesSelected() {
+        val sitePermissions = _viewState.value.sitesPermissionsAllowed.toMutableList()
+        val locationPermissions = _viewState.value.locationPermissionsAllowed.toMutableList()
+        viewModelScope.launch(dispatcherProvider.io()) {
+            geolocationPermissions.clearAll()
```

Comment:
do we have plans to unify locations with the rest of permissions? Asking because I wonder if this logic should belong to a repository rather than to the view model.

Reply:
cool, are you keeping track of all these follow-up tasks in Asana?